Error parsing WhatsApp DB (Comparison method violates its general contract)

Another user reported the following error (it happens both with 4.1.x and master).
Analysing the database, the issue is caused by the way messages are sorted.
It may fail when a very specific combination of records (with zeroes or null values in the columns used) is present, which is the case of the triggering database.
I will submit a fix shortly.

```
2024-10-15 09:32:48	[ERROR]	[parsers.whatsapp.WhatsAppParser]			Error parsing WhatsApp: Item: XXX/msgstore.db type: sqlite size: 61648896
org.apache.tika.exception.TikaException: WAExtractorException Exception
	at iped.parsers.whatsapp.WhatsAppParser.parseWhatsappMessages(WhatsAppParser.java:399) ~[iped-parsers-impl-4.2-snapshot.jar:?]
	at iped.parsers.whatsapp.WhatsAppParser.parse(WhatsAppParser.java:259) [iped-parsers-impl-4.2-snapshot.jar:?]
	at org.apache.tika.parser.CompositeParser.parse(CompositeParser.java:298) [tika-core-2.4.0-p1.jar:2.4.0]
	at iped.parsers.standard.StandardParser.parse(StandardParser.java:245) [iped-parsers-impl-4.2-snapshot.jar:?]
	at iped.engine.io.ParsingReader$BackgroundParsing.run(ParsingReader.java:247) [iped-engine-4.2-snapshot.jar:?]
	at java.util.concurrent.Executors$RunnableAdapter.call(Unknown Source) [?:?]
	at java.util.concurrent.FutureTask.run(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) [?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) [?:?]
	at java.lang.Thread.run(Unknown Source) [?:?]
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.util.ComparableTimSort.mergeHi(Unknown Source) ~[?:?]
	at java.util.ComparableTimSort.mergeAt(Unknown Source) ~[?:?]
	at java.util.ComparableTimSort.mergeForceCollapse(Unknown Source) ~[?:?]
	at java.util.ComparableTimSort.sort(Unknown Source) ~[?:?]
	at java.util.Arrays.sort(Unknown Source) ~[?:?]
	at java.util.Arrays.sort(Unknown Source) ~[?:?]
	at java.util.ArrayList.sort(Unknown Source) ~[?:?]
	at java.util.Collections.sort(Unknown Source) ~[?:?]
	at iped.parsers.whatsapp.ExtractorAndroidNew.extractChatList(ExtractorAndroidNew.java:150) ~[iped-parsers-impl-4.2-snapshot.jar:?]
	at iped.parsers.whatsapp.Extractor.getChatList(Extractor.java:34) ~[iped-parsers-impl-4.2-snapshot.jar:?]
	at iped.parsers.whatsapp.WhatsAppParser.parseWhatsappMessages(WhatsAppParser.java:391) ~[iped-parsers-impl-4.2-snapshot.jar:?]
	... 9 more

```

--- a/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
@@ -637,7 +637,7 @@ public class Message implements Comparable<Message> {
     }
 
     public long getSortId() {
-        return sortId == 0 ? id : sortId;
+        return sortId == 0 ? timeStamp == null ? 0 : timeStamp.getTime() : sortId;
     }
 
     public void setSortId(long sortId) {
@@ -690,14 +690,6 @@ public class Message implements Comparable<Message> {
 
     @Override
     public int compareTo(Message o) {
-        int comp = Long.compare(getSortId(), o.getSortId());
-        if (comp != 0) {
-            return comp;
-        }
-        return Long.compare(getTime(), o.getTime());
-    }
-
-    private long getTime() {
-        return timeStamp == null ? 0 : timeStamp.getTime();
+        return Long.compare(getSortId(), o.getSortId());
     }
 }

--- a/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
@@ -637,7 +637,7 @@ public class Message implements Comparable<Message> {
     }
 
     public long getSortId() {
-        return sortId == 0 ? timeStamp == null ? 0 : timeStamp.getTime() : sortId;
+        return sortId == 0 ? timeStamp == null ? 0 : timeStamp.getTime() : id;
     }
 
     public void setSortId(long sortId) {

--- a/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
@@ -637,7 +637,7 @@ public class Message implements Comparable<Message> {
     }
 
     public long getSortId() {
-        return sortId;
+        return sortId == 0 ? id : sortId;
     }
 
     public void setSortId(long sortId) {
@@ -690,18 +690,14 @@ public class Message implements Comparable<Message> {
 
     @Override
     public int compareTo(Message o) {
-        if (getSortId() != 0 && o.getSortId() != 0) {
-            int comp = Long.compare(getSortId(), o.getSortId());
-            if (comp != 0) {
-                return comp;
-            }
+        int comp = Long.compare(getSortId(), o.getSortId());
+        if (comp != 0) {
+            return comp;
         }
-        if (getTimeStamp() != null && o.getTimeStamp() != null) {
-            int comp = getTimeStamp().compareTo(o.getTimeStamp());
-            if (comp != 0) {
-                return comp;
-            }
-        }
-        return Long.compare(getId(), o.getId());
+        return Long.compare(getTime(), o.getTime());
+    }
+
+    private long getTime() {
+        return timeStamp == null ? 0 : timeStamp.getTime();
     }
 }

--- a/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
+++ b/iped-parsers/iped-parsers-impl/src/main/java/iped/parsers/whatsapp/Message.java
@@ -637,7 +637,7 @@ public class Message implements Comparable<Message> {
     }
 
     public long getSortId() {
-        return sortId == 0 ? timeStamp == null ? 0 : timeStamp.getTime() : id;
+        return sortId != 0 ? sortId : timeStamp != null ? timeStamp.getTime() : id;
     }
 
     public void setSortId(long sortId) {

